File: Back/PORTAIL_RH/src/main/java/com/example/PORTAIL_RH/feed_service/Reaction_service/Service/ReactionService.java
package com.example.PORTAIL_RH.feed_service.Reaction_service.Service;

import com.example.PORTAIL_RH.feed_service.Reaction_service.DTO.CommentDTO;
import com.example.PORTAIL_RH.feed_service.Reaction_service.DTO.CommentRequest;
import com.example.PORTAIL_RH.feed_service.Reaction_service.DTO.ReactionDTO;
import com.example.PORTAIL_RH.feed_service.Reaction_service.DTO.ReactionRequest;
import com.example.PORTAIL_RH.feed_service.Reaction_service.DTO.ReactionSummaryDTO;
import com.example.PORTAIL_RH.feed_service.Reaction_service.DTO.IdeaRatingDTO;
import com.example.PORTAIL_RH.feed_service.Reaction_service.DTO.IdeaRatingRequest;
import com.example.PORTAIL_RH.feed_service.Reaction_service.DTO.CommentUpdateRequest;

import java.util.List;

public interface ReactionService {
    ReactionDTO createOrUpdateReaction(ReactionRequest reactionRequest);
    List<ReactionDTO> getReactionsByPublicationId(Long publicationId);
    ReactionSummaryDTO getReactionSummaryByPublicationId(Long publicationId);
    void deleteReaction(Long userId, Long publicationId);
    CommentDTO createComment(CommentRequest commentRequest);
    CommentDTO updateComment(Long commentId, CommentUpdateRequest updateRequest);
    List<CommentDTO> getCommentsByPublicationId(Long publicationId);
    void deleteComment(Long commentId, Long userId);
    IdeaRatingDTO createIdeaRating(IdeaRatingRequest ratingRequest);
    List<IdeaRatingDTO> getIdeaRatingsByPublicationId(Long publicationId);
    void deleteIdeaRating(Long userId, Long publicationId);
}

File: Front/PORTAIL-RH/src/app/services/reaction.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpErrorResponse, HttpHeaders } from '@angular/common/http';
import { Observable, throwError } from 'rxjs';
import { catchError } from 'rxjs/operators';

export interface ReactionDTO {
    id?: number;
    userId: number;
    userNom: string;
    userPrenom: string;
    publicationId: number;
}

export interface CommentDTO {
    id?: number;
    userId: number;
    userNom: string;
    userPrenom: string;
    userPhoto: string | null;
    publicationId: number;
    content: string;
    createdAt: Date;
}

export interface ReactionRequest {
    userId: number;
    publicationId: number;
}

export interface ReactionSummaryDTO {
    publicationId: number;
    totalLikes: number;
}

@Injectable({
    providedIn: 'root'
})
export class ReactionService {
    private apiUrl = 'http://localhost:8080/api/reactions';

    constructor(private http: HttpClient) {}

    // Helper method to create headers for JSON requests
    private getJsonHeaders(): HttpHeaders {
        return new HttpHeaders({
            'Content-Type': 'application/json'
        });
    }

    // Create or update a like on a publication
    createOrUpdateReaction(reactionRequest: ReactionRequest): Observable<ReactionDTO> {
        return this.http.post<ReactionDTO>(`${this.apiUrl}/like`, reactionRequest, {
            headers: this.getJsonHeaders()
        });
    }

    // Fetch all likes for a publication
    getReactionsByPublicationId(publicationId: number): Observable<ReactionDTO[]> {
        return this.http.get<ReactionDTO[]>(`${this.apiUrl}/publication/${publicationId}/likes`);
    }

    // Fetch the reaction summary (total likes) for a publication
    getReactionSummaryByPublicationId(publicationId: number): Observable<ReactionSummaryDTO> {
        return this.http.get<ReactionSummaryDTO>(`${this.apiUrl}/publication/${publicationId}/summary`);
    }

    // Delete a like for a user and publication
    deleteReaction(userId: number, publicationId: number): Observable<void> {
        return this.http.delete<void>(`${this.apiUrl}/user/${userId}/publication/${publicationId}/like`);
    }

    // Delete an idea rating for a user and publication
    deleteIdeaRating(userId: number, publicationId: number): Observable<void> {
        return this.http.delete<void>(`${this.apiUrl}/user/${userId}/publication/${publicationId}/rating`).pipe(
            catchError(this.handleError)
        );
    }

    // Fetch all comments for a publication (optional, prefer PublicationService)
    getCommentsByPublicationId(publicationId: number): Observable<CommentDTO[]> {
        return this.http.get<CommentDTO[]>(`${this.apiUrl}/publication/${publicationId}/comments`);
    }

    // Delete a comment
    deleteComment(commentId: number, userId: number): Observable<void> {
        return this.http.delete<void>(`${this.apiUrl}/comment/${commentId}/user/${userId}`);
    }

    private handleError(error: HttpErrorResponse): Observable<never> {
        let errorMessage = 'Une erreur est survenue.';
        if (error.error instanceof ErrorEvent) {
            // Client-side error
            errorMessage = `Erreur côté client : ${error.error.message}`;
        } else {
            // Server-side error
            if (error.status === 404) {
                errorMessage = 'Ressource introuvable. Vérifiez l\'URL ou l\'existence de la ressource.';
            } else if (error.status === 400) {
                errorMessage = error.error?.message || 'Requête invalide. Vérifiez les données envoyées.';
            } else if (error.status === 500) {
                errorMessage = 'Erreur interne du serveur. Contactez l\'administrateur.';
            } else {
                errorMessage = `Code d'erreur ${error.status}: ${error.error?.message || error.message}`;
            }
        }
        console.error(errorMessage, error);
        return throwError(() => new Error(errorMessage));
    }
}